src/checkers/inference/solver/backend/encoder/ConstraintEncoderCoordinator.java

```diff
@@ -39,10 +39,9 @@
                 return encoder.encodeVariable_Constant((VariableSlot) constraint.getFirst(), (ConstantSlot) constraint.getSecond());
             case CONSTANT_VARIABLE:
                 return encoder.encodeConstant_Variable((ConstantSlot) constraint.getFirst(), (VariableSlot) constraint.getSecond());
-            case CONSTANT_CONSTANT:
```

Comment:
Removing this branch makes assumption that `ConstraintManager` never allow this case to be reached. But it's possible that during refactoring this case is reached someday. Either error abort here or leave it as it is. The reason for the latter option is to keep inference framework classes and solver classes independent and doesn't correlate with each other by assumptions that may break someday.

Reply:
Will issue error.

> keep inference framework classes and solver classes independent 

Dragging the qualifier hierarchy all the way down to the solver via the constraint verifier is not independent from my point of view, since front end code defining subtyping will affect solver code.

This PR actually does make the two more independent, because the normalization occurs before the constraints are added to the set, and the back end solvers can perform blind serialization without needing to check constant-constant cases and be "smart" again.